### ISSUE TITLE
Implement QLoRA

### ISSUE THREAD

**Issue by sr5434** [stanford-crfm/levanter#508]
I am trying to fine-tune Phind 34B using Levanter on a TPU v4-8 VM. How can I implement QLoRA without getting an error telling me I must use full or half precision?

**Comment by dlwh** [comment]
Yeah sorry, we don't have QLoRA. I don't really know the best way to do it in JAX is and I'm unlikely to implement it. Happy to give advice.

I think a v4-8 should be able to fit 34B for unquantized LoRA with mixed fp32/bf16 though?

**Comment by sr5434** [comment]
I think it is trying to run a distributed training, and duplicating it across each core. Is there a way for it to use the TPU-v4 as one big device? Also, could I implement QLoRA by just doing: 

```python
weights.to(jnp.int4)
```

**Comment by dlwh** [comment]
We shard the model parameters so each should only get 1/4 of the model... If it's not fitting it's a bug I'm pretty sure.

Quantization is more complicated than that, unfortunately. There are different strategies and I don't know the best one on TPU/JAX.

**Comment by sr5434** [comment]
I think it might be a bug, as I'm getting an OOM. However, it is not a TPU OOM, but a CPU one for some reason.

**Comment by sr5434** [comment]
I'm using a virtual environment for python 3.10. [This gist](https://gist.github.com/sr5434/b5d75b6b6e527d9551a3d495552aa120) contains my script and config. Also, I can't tell if the error message is about the CPU running out of ram(implied by the tmalloc warnings) or the TPU, implied by the error.
These are my logs:

```
2024-03-05T22:15:58 - 0 - levanter.distributed - distributed.py:252 - INFO :: Successfully started ray head on port 61964.
2024-03-05T22:15:58 - 0 - levanter.distributed - distributed.py:267 - INFO :: ray.init(address='10.130.0.8:61964', namespace='levanter', **{})
2024-03-05 22:15:58,436 INFO worker.py:1540 -- Connecting to existing Ray cluster at address: 10.130.0.8:61964...
2024-03-05 22:15:58,445 INFO worker.py:1715 -- Connected to Ray cluster. View the dashboard at 10.130.0.8:8265 
You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama.LlamaTokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
2024-03-05T22:16:01 - 0 - levanter.data.shard_cache - shard_cache.py:1575 - INFO :: Loading cache from ./cache
2024-03-05T22:16:01 - 0 - levanter.data.shard_cache - shard_cache.py:97 - INFO :: Cache already finished. Skipping.
2024-03-05T22:16:01 - 0 - levanter.tracker.wandb - wandb.py:189 - INFO :: Setting wandb code_dir to /home/samir_rangwalla/levanter
2024-03-05T22:16:01 - 0 - levanter.tracker.wandb - wandb.py:189 - INFO :: Setting wandb code_dir to /home/samir_rangwalla/levanter
train:   0%|                                                                                                                                                  | 0/200 [00:00<?, ?it/s]2024-03-05T22:16:02 - 0 - __main__ - gsm8k_lora.py:156 - INFO :: Loading pretrained model from Phind/Phind-CodeLlama-34B-v2
/home/samir_rangwalla/levanter/py_venv/lib/python3.10/site-packages/haliax/random.py:35: FutureWarning: normal accepts a single key, but was given a key array of shape (1, 2) != (). Use jax.vmap for batching. In a future JAX version, this will be an error.
  jax_array = jrandom.normal(key=key, shape=jax_shape, dtype=dtype)
/home/samir_rangwalla/levanter/py_venv/lib/python3.10/site-packages/jax/_src/random.py:740: FutureWarning: uniform accepts a single key, but was given a key array of shape (1,) != (). Use jax.vmap for batching. In a future JAX version, this will be an error.
  u = uniform(key, shape, dtype, lo, hi)  # type: ignore[arg-type]
tcmalloc: large alloc 17314086912 bytes == 0x13dad98000 @  0x7f26cb2b7680 0x7f26cb2d8824 0x7f26c9cb1b14 0x7f26c9cb22af 0x7f26c9cd25ce 0x7f26c9d24726 0x7f26c9d24f8f 0x7f26c9d2515b 0x53a6a4 0x7f26c9cbae8e 0x5aa585 0x629bc0 0x7f26c9cbae8e 0x5aa585 0x629bc0 0x5aa585 0x629bc0 0x62989a 0x5abdcb 0x629bc0 0x62c709 0x62c83a 0x6e9cc7 0x62ab94 0x5aea04 0x629bc0 0x5a94cb 0x629bc0 0x5a94cb 0x54858a 0x5adeb3
tcmalloc: large alloc 6442450944 bytes == 0x17e3d98000 @  0x7f26cb2b7680 0x7f26cb2d8824 0x7f26c9cb1b14 0x7f26c9cb22af 0x7f26c9cd25ce 0x7f26c9d24726 0x7f26c9d24f8f 0x7f26c9d2515b 0x53a6a4 0x7f26c9cbae8e 0x5aa585 0x629bc0 0x7f26c9cbae8e 0x5aa585 0x629bc0 0x5aa585 0x629bc0 0x62989a 0x5abdcb 0x629bc0 0x62c709 0x62c83a 0x6e9cc7 0x62ab94 0x5aea04 0x629bc0 0x5a94cb 0x629bc0 0x5a94cb 0x54858a 0x5adeb3
...pattern continues...
/home/samir_rangwalla/levanter/py_venv/lib/python3.10/site-packages/jax/_src/interpreters/mlir.py:914: UserWarning: Some donated buffers were not usable: ShapedArray(float32[48,8,2048]), ShapedArray(float32[48,8,8,128,8]), ShapedArray(float32[48,8,2048]), ShapedArray(float32[48,8,128,8]), ShapedArray(float32[48,8,2048]), ShapedArray(float32[48,8,128,8]), ShapedArray(float32[48,8,64,128]), ShapedArray(float32[48,2048,8]), ShapedArray(float32[48,8,2048]), ShapedArray(float32[48,22016,8]), ShapedArray(float32[48,8,2048]), ShapedArray(float32[48,22016,8]), ShapedArray(float32[48,8,22016]), ShapedArray(float32[48,2048,8]), ShapedArray(float32[8,2048]), ShapedArray(float32[32000,8]).
See an explanation at https://jax.readthedocs.io/en/latest/faq.html#buffer-donation.
  warnings.warn("Some donated buffers were not usable:"
2024-03-05T22:20:05 - 0 - __main__ - gsm8k_lora.py:183 - INFO :: Total parameter count: 33798752256
2024-03-05T22:20:05 - 0 - __main__ - gsm8k_lora.py:184 - INFO :: Trainable parameter count: 54781952
2024-03-05T22:20:05 - 0 - __main__ - gsm8k_lora.py:185 - INFO :: Fraction of parameters that are trainable: 0.0016208276443185868
You're using a LlamaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
...Error traceback...
jaxlib.xla_extension.XlaRuntimeError: RESOURCE_EXHAUSTED: Allocation (size=51539607552) would exceed memory (size=34359738368) :: #allocation478 [shape = 'bf16[48,8,8192,8192]{2,3,1,0:T(8,128)(2,1)}', space=hbm, size = 0xffffffffffffffff, tag = 'output of broadcast.672@{}'] :: <no-hlo-instruction>
```

**Comment by dlwh** [comment]
ok that one is easy I think. decrease the `trainer.train_batch_size` (probably defaulting to 64?) to something smaller.

**Comment by sr5434** [comment]
My batch size is 32. Is that also too high?

**Comment by dlwh** [comment]
probably for this big of a model with that long of sequence length...

**Comment by sr5434** [comment]
Alright, thanks. I will try rerunning the job later and let you know how it goes.

**Comment by dlwh** [comment]
if you're using the alpaca  or gsm8k script, you can also reduce the sequence length with max_tune_length if that works for your task...

**Comment by sr5434** [comment]
I ended up increasing the length; as I am finetuning it on code and code can(in theory) get quite long.

**Comment by sr5434** [comment]
Ok, now I have tested the script with a lower batch size, but still get OOM. I was unable to get a batch size lower than 4 because of a modulo by zero error. It is slightly different, returning a "TPU permanent compile error", with the process finally being killed by a broken pipe error.

**Comment by dlwh** [comment]
hrm ok, maybe it's not using flash attention (are you using a recent fork of main?) can you print `model_config.use_flash_attention`?

**Comment by sr5434** [comment]
I believe I installed the PyPi package

**Comment by dlwh** [comment]
Ok try main and see if that fixes it.

**Comment by sr5434** [comment]
Alright, I can try it later and let you know

**Comment by sr5434** [comment]
I just double checked and I did in fact build it from source(about 5 days ago). According to the model config, it is using flash attention.

**Comment by dlwh** [comment]
ok cool, Thanks for double checking. i will take a look sometime in the
next few days.  If FA is on it really shouldn't be making that large of a
tensor, which is weird.

**Comment by sr5434** [comment]
Ok, sounds great

**Comment by sr5434** [comment]
Are there any updates in the issue?

**Comment by sr5434** [comment]
No problem!

**Comment by dlwh** [comment]
Think I've about got it. Couple of missteps but I'm almost there I think

**Comment by dlwh** [comment]
should be fixed after #523 . Can you give that a try? max batch size is 4 on a v4-8 (with seq len 8192, which is just super long)

```
2024-03-12T19:05:00 - 0 - __main__ - lora_code.py:156 - INFO :: Loading pretrained model from Phind/Phind-CodeLlama-34B-v2
Loading weights: 100%|█████████████████████████████████████████████| 59/59 [00:08<00:00,  7.04it/s]
Loading weights: 100%|█████████████████████████████████████████████| 63/63 [00:09<00:00,  6.77it/s]
Loading weights: 100%|█████████████████████████████████████████████| 61/61 [00:09<00:00,  6.11it/s]
Loading weights: 100%|█████████████████████████████████████████████| 63/63 [00:08<00:00,  7.10it/s]
Loading weights: 100%|█████████████████████████████████████████████| 63/63 [00:08<00:00,  7.00it/s]
Loading weights: 100%|█████████████████████████████████████████████| 63/63 [00:08<00:00,  7.06it/s]
Loading weights: 100%|█████████████████████████████████████████████| 63/63 [00:08<00:00,  7.21it/s]
before loraizes: 100%|█████████████████████████████████████████████| 63/63 [00:08<00:00,  6.59it/s]
/home/dlwh/venv310/lib/python3.10/site-packages/jax/_src/interpreters/mlir.py:772: UserWarning: Some donated buffers were not usable: ShapedArray(float32[48,1,2048]), ShapedArray(float32[48,8,8,128,1]), ShapedArray(float32[48,1,2048]), ShapedArray(float32[48,8,128,1]), ShapedArray(float32[48,1,2048]), ShapedArray(float32[48,8,128,1]), ShapedArray(float32[48,1,64,128]), ShapedArray(float32[48,2048,1]), ShapedArray(float32[48,1,2048]), ShapedArray(float32[48,22016,1]), ShapedArray(float32[48,1,2048]), ShapedArray(float32[48,22016,1]), ShapedArray(float32[48,1,22016]), ShapedArray(float32[48,2048,1]), ShapedArray(float32[1,2048]), ShapedArray(float32[32000,1]).
See an explanation at https://jax.readthedocs.io/en/latest/faq.html#buffer-donation.
  warnings.warn("Some donated buffers were not usable:"
2024-03-12T19:08:57 - 0 - __main__ - lora_code.py:192 - INFO :: Total parameter count: 33750818048
2024-03-12T19:08:57 - 0 - __main__ - lora_code.py:193 - INFO :: Trainable parameter count: 6847744
You're using a LlamaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
train:   2%|█                                         | 5/200 [05:24<1:53:47, 35.01s/it, loss=1.48]
```

**Comment by sr5434** [comment]
Thank you so much @dlwh! I was able to fine tune the model last night. I love the work you are doing with this repo!
